bundles/org.openhab.binding.caddx/src/main/java/org/openhab/binding/caddx/internal/handler/CaddxBridgeHandler.java

```diff
@@ -0,0 +1,408 @@
+        return discoveryService;
+    }
+
+    public void setDiscoveryService(CaddxDiscoveryService discoveryService) {
+        this.discoveryService = discoveryService;
+    }
+
+    /**
+     * Constructor.
+     *
+     * @param bridge
+     */
+    public CaddxBridgeHandler(SerialPortManager portManager, Bridge bridge) {
+        super(bridge);
+
+        this.portManager = portManager;
+    }
+
+    private void init() {
+        CaddxBridgeConfiguration configuration = getConfigAs(CaddxBridgeConfiguration.class);
+
+        protocol = configuration.getProtocol();
+        serialPortName = configuration.getSerialPort();
+        baudRate = configuration.getBaudrate().intValue();
+        updateStatus(ThingStatus.OFFLINE);
+
+        // create & start panel interface
+        logger.info("starting interface at port {} with baudrate {}", serialPortName, baudRate);
+
+        try {
+            communicator = new CaddxCommunicator(portManager, protocol, serialPortName, baudRate);
+        } catch (IOException | TooManyListenersException | UnsupportedCommOperationException | PortInUseException e) {
+            logger.warn("Cannot initialize Communication.", e);
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                    "Communication cannot be initialized");
+
+            throw new IllegalArgumentException();
+        }
+
+        CaddxCommunicator comm = communicator;
+        if (comm != null) {
+            comm.addListener(this);
+
+            // Send discovery commands for the things
+            comm.transmit(new CaddxMessage(DISCOVERY_ZONES_SNAPSHOT_REQUEST_00, false));
+            comm.transmit(new CaddxMessage(DISCOVERY_ZONES_SNAPSHOT_REQUEST_10, false));
+            comm.transmit(new CaddxMessage(DISCOVERY_ZONES_SNAPSHOT_REQUEST_20, false));
+            comm.transmit(new CaddxMessage(DISCOVERY_PARTITION_STATUS_REQUEST_0, false));
+            comm.transmit(new CaddxMessage(DISCOVERY_PARTITIONS_SNAPSHOT_REQUEST, false));
+        }
+        // list all channels
+        if (logger.isTraceEnabled()) {
+            logger.trace("list all {} channels:", getThing().getChannels().size());
+            for (Channel c : getThing().getChannels()) {
+                logger.trace("Channel Type {} UID {}", c.getChannelTypeUID(), c.getUID());
+            }
+        }
+    }
+
+    @Override
+    public void initialize() {
+        logger.trace("Initializing the Bridge handler.");
+
+        init();
+    }
+
+    @Override
+    public void dispose() {
+        CaddxCommunicator comm = communicator;
+        if (comm != null) {
+            comm.stop();
+            comm = null;
+        }
+
+        if (discoveryService != null) {
+            unregisterDiscoveryService();
+        }
+
+        super.dispose();
+    }
+
+    public @Nullable Thing findThing(CaddxThingType caddxThingType, @Nullable Integer partition, @Nullable Integer zone,
```

Comment:
This is called multiple times every time an event is received. You should make sure this method is a fast as possible or it would easily become a performance bottleneck in this binding. Not to mention that openhab often runs on systems with limited cpu resources, try not to be wasteful.

Reply:
It is called multiple times only when receiving specific messages. These messages are received during discovery when the serial communication is initialized.
On normal operation it is called only once per event.

I cannot think of a way to make it faster. Any suggestion?

Reply:
You could override the `childHandlerInitialized` and `childHandlerDisposed` methods so that you can categorize the handler and things into their CaddxThingType types. 
For the PARTITION and ZONE types you could get away with using an array of things/handlers that you could then do easy lookups against by simply checking if the thing/handler at a particular array index is present.
For keypads you would likely use a map and use the keypad address as the keys.

In general that is the idea. You want this method to make lookups in constant time rather than scale with the number of things on the bridge.

Reply:
Done